src/drawer.tsx

```diff
@@ -0,0 +1,306 @@
+import {IconButton} from './icon-button';
+import IconCloseRegular from './generated/mistica-icons/icon-close-regular';
+import Box from './box';
+import * as styles from './drawer.css';
+import classnames from 'classnames';
+import {Portal} from './portal';
+import {useIsInViewport, useScreenSize, useTheme} from './hooks';
+import FocusTrap from './focus-trap';
+import {useSetModalStateEffect} from './modal-context-provider';
+import ButtonLayout from './button-layout';
+import {ButtonLink, ButtonPrimary, ButtonSecondary} from './button';
+import Divider from './divider';
+import {getPrefixedDataAttributes} from './utils/dom';
+import * as tokens from './text-tokens';
+
+import type {DataAttributes, HeadingType, TrackingEvent} from './utils/types';
+
+const PADDING_X_DESKTOP = 40;
+const PADDING_X_MOBILE = 16;
+const PADDING_X_TABLET = 16;
+const WIDTH_CONTENT = 388;
+const WIDTH_DESKTOP = WIDTH_CONTENT + PADDING_X_DESKTOP * 2;
+const WIDTH_TABLET = WIDTH_CONTENT + PADDING_X_TABLET * 2;
+
+/**
+ * Renders divider or a div with transparent border to avoid the small but noticeable layout shift on scroll
+ */
+const MaybeDivider = ({show}: {show: boolean}) =>
+    show ? <Divider /> : <div style={{borderBottom: '1px solid transparent'}} />;
+
+/**
+ * Restores the focus to the element that was focused before the Drawer was opened
+ */
+const useRestoreFocus = () => {
+    const activeElementRef = React.useRef<HTMLElement | null>(document.activeElement as HTMLElement);
+    React.useEffect(() => {
+        const elementToFocus = activeElementRef.current;
+        return () => {
+            elementToFocus?.focus?.();
+        };
+    }, []);
+};
+
+type DrawerLayoutProps = {
+    width?: number;
+    children: React.ReactNode;
+    onClose: () => void;
+    onDismiss?: () => void;
+};
+
+type DrawerPropsRef = {
+    close: () => Promise<void>;
+    dismiss: () => Promise<void>;
+};
+
+const DrawerLayout = React.forwardRef<DrawerPropsRef, DrawerLayoutProps>(
+    ({width, children, onClose, onDismiss}, ref) => {
+        useSetModalStateEffect();
+        useRestoreFocus();
+        const {isMobile, isTablet} = useScreenSize();
+        const [isOpen, setIsOpen] = React.useState(false);
+        const widthStyle = isMobile ? 'auto' : width || (isTablet ? WIDTH_TABLET : WIDTH_DESKTOP);
+
+        const open = React.useCallback((node: HTMLDivElement) => {
+            if (node) {
+                // small delay to allow the Portal to be mounted
+                setTimeout(() => {
+                    setIsOpen(true);
+                }, 50);
+            }
+        }, []);
+
+        const close = React.useCallback(() => {
+            setIsOpen(false);
+            return new Promise((resolve) => {
+                setTimeout(resolve, styles.ANIMATION_DURATION_MS);
+            }).then(onClose);
+        }, [onClose]);
+
+        const dismiss = React.useCallback(() => {
+            return close().then(() => onDismiss?.());
```

Comment:
what's the difference between `onDismiss` and `onClose`? I initially thought that `onDismiss` was called immediately and `onClose` when the animation ends, but it seems both are called when the animation ends :/

Reply:
ok, I see the difference. `onDismiss` is only called when the drawer is closed by a dismiss action (esc key /close button, click outside). But can't we call `onDismiss` immediately without waiting for animation?

Reply:
Yes, we could call it immediately but I decided to not emit the close/dismiss event while the drawer is visible.